pkg/kubelet/kuberuntime/kuberuntime_container_linux.go

```diff
@@ -107,17 +107,22 @@ func (m *kubeGenericRuntimeManager) generateLinuxContainerResources(pod *v1.Pod,
 
 	lcr.HugepageLimits = GetHugepageLimitsFromResources(container.Resources)
 
-	if swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo); utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
-		// NOTE(ehashman): Behaviour is defined in the opencontainers runtime spec:
-		// https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#memory
-		switch m.memorySwapBehavior {
-		case kubelettypes.LimitedSwap:
-			swapConfigurationHelper.ConfigureLimitedSwap(lcr, pod, container)
-		default:
-			swapConfigurationHelper.ConfigureUnlimitedSwap(lcr)
+	// swap is only configured if a swap controller is available.
+	if swapControllerAvailable() {
+		if swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo); utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
+			// NOTE(ehashman): Behaviour is defined in the opencontainers runtime spec:
+			// https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#memory
+			switch m.memorySwapBehavior {
+			case kubelettypes.LimitedSwap:
+				swapConfigurationHelper.ConfigureLimitedSwap(lcr, pod, container)
+			default:
+				swapConfigurationHelper.ConfigureUnlimitedSwap(lcr)
+			}
+		} else {
+			swapConfigurationHelper.ConfigureNoSwap(lcr)
 		}
-	} else {
-		swapConfigurationHelper.ConfigureNoSwap(lcr)
+	} else if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
+		klog.ErrorS(errors.New("no cgroup swap controller present"), "ignoring NodeSwap feature", "pod", klog.KObj(pod), "containerName", container.Name)
```

Comment:
Try to write the error logging as if the feature is already stable. The value of the feature gate doesn't seem very relevant here.

Reply:
Thanks @sftim. I elected to always log (at `InfoS`) that we're not configuring swap but to include the `SwapBehavior` value as well. This should be enough to give users a signal that its being skipped without making the logic in the function overly complex. Hope this is sufficient from your side.